src/bgw/scheduler.c

```diff
@@ -0,0 +1,395 @@
+}
+
+static
+List *
+scheduled_bgw_job_list_create(MemoryContext mctx)
+{
+	List	   *res = NIL;
+	List	   *jobs = ts_bgw_job_get_all(mctx);
+	ListCell   *lc;
+	MemoryContext old;
+
+	foreach(lc, jobs)
+	{
+		BgwJob	   *job = lfirst(lc);
+		ScheduledBgwJob *sjob = scheduled_bgw_job_create(job, mctx);
+
+		old = MemoryContextSwitchTo(mctx);
+		res = lappend(res, sjob);
+		MemoryContextSwitchTo(old);
+	}
+	return res;
+}
+
+static
+bool
+scheduled_bgw_job_list_start_workers()
+{
+	ListCell   *lc;
+	bool		running = false;
+
+	foreach(lc, scheduled_jobs)
+	{
+		ScheduledBgwJob *sjob = lfirst(lc);
+
+		if (scheduled_bgw_job_is_running(sjob))
+		{
+			running = true;
+			continue;
+		}
+		if (!sjob->enabled || get_timeout(sjob->next_start) > 0)
+			continue;
+		scheduled_bgw_job_start_worker(sjob);
+		if (sjob->handle != NULL)
+			running = true;
+	}
+	return running;
+}
+
+static TimestampTz
+scheduled_bgw_job_list_next_start(TimestampTz next_start)
+{
+	ListCell   *lc;
+
+	foreach(lc, scheduled_jobs)
+	{
+		ScheduledBgwJob *sjob = lfirst(lc);
+
+		if (!sjob->enabled || scheduled_bgw_job_is_running(sjob))
+			continue;
+		if (sjob->next_start < next_start)
+			next_start = sjob->next_start;
+	}
+	return next_start;
+}
+
+static TimestampTz
+scheduled_bgw_job_list_wake_for_timeout(TimestampTz next_start)
+{
+	ListCell   *lc;
+
+	foreach(lc, scheduled_jobs)
+	{
+		ScheduledBgwJob *sjob = lfirst(lc);
+
+		if (!scheduled_bgw_job_is_running(sjob))
+			continue;
+
+		if (sjob->timeout_at < next_start)
+			next_start = sjob->timeout_at;
+	}
+	return next_start;
+}
+
+
+
+
+static void
+scheduled_bgw_job_list_terminate_all()
+{
+	ListCell   *lc;
+
+	foreach(lc, scheduled_jobs)
+	{
+		ScheduledBgwJob *sjob = lfirst(lc);
+
+		if (scheduled_bgw_job_is_running(sjob))
+			TerminateBackgroundWorker(sjob->handle);
+	}
+}
+
+static void
+scheduled_bgw_job_list_check_stopped_or_timed_out_jobs()
+{
+	ListCell   *lc;
+
+	foreach(lc, scheduled_jobs)
+	{
+		BgwHandleStatus status;
+		pid_t		pid;
+		ScheduledBgwJob *sjob = lfirst(lc);
+		TimestampTz now = GetCurrentTimestamp();
+
+		if (sjob->handle == NULL)
+			continue;
+
+		status = GetBackgroundWorkerPid(sjob->handle, &pid);
+
+		switch (status)
+		{
+			case BGWH_POSTMASTER_DIED:
+				bgw_scheduler_on_postmaster_death();
+				break;
+			case BGWH_NOT_YET_STARTED:
+				elog(ERROR, "unexpected bgworker state %d", status);
+				break;
+			case BGWH_STARTED:
+				/* still running */
+				if (now > sjob->timeout_at)
+				{
+					elog(WARNING, "terminating background worker \"%s\" due to timeout", NameStr(sjob->job->fd.job_name));
+					TerminateBackgroundWorker(sjob->handle);
+
+					/*
+					 * reset timeout at so we don't continously send
+					 * terminates
+					 */
+					sjob->timeout_at = DT_NOEND;
+				}
+				break;
+			case BGWH_STOPPED:
+				StartTransactionCommand();
+				scheduled_bgw_job_mark_stopped(sjob);
+				CommitTransactionCommand();
+				break;
+		}
+	}
+}
+
+void
+ts_bgw_scheduler_process(int32 ttl_ms)
```

Comment:
Why does the scheduler take a TTL? What's a potential use-case for this? Shouldn't the scheduler just run forever until some signal?

Reply:
This is purely for making it easier to test. I'll add documentation.